### ISSUE TITLE
[Query] Options for multi value parameters?

### ISSUE THREAD

**Issue by colemars** [reduxjs/redux-toolkit#1469]
Hello 👋 
      
Currently, `fetchBaseQuery` utilizes `UrlSearchParams` for preparing its query params. `UrlSearchParams` will parse multi value params (e.g. `param=[1,2,3]`) as a comma separated list, `...&param=1,2,3`.

However,  it's not uncommon when working with various 3rd parties to encounter different requirements. 

Most notably, AWS API Gateway [looks for repeating keys](https://aws.amazon.com/blogs/compute/support-for-multi-value-parameters-in-amazon-api-gateway/) (`...&param=1&param=2&param=3`).

Would exposing some sort of config here to support other multi value parameter strategies be valid?

as a minimal, hasty example:

```diff
diff --git a/node_modules/@reduxjs/toolkit/src/query/fetchBaseQuery.ts b/node_modules/@reduxjs/toolkit/src/query/fetchBaseQuery.ts
index 49605fc..b9222a6 100644
--- a/node_modules/@reduxjs/toolkit/src/query/fetchBaseQuery.ts
+++ b/node_modules/@reduxjs/toolkit/src/query/fetchBaseQuery.ts
@@ -22,6 +22,7 @@ type CustomRequestInit = Override<
 export interface FetchArgs extends CustomRequestInit {
   url: string
   params?: Record<string, any>
+  paramsOpts?: { arrayFormat: 'default' | 'repeat' }
   body?: any
   responseHandler?: ResponseHandler
   validateStatus?: (response: Response, body: any) => boolean
@@ -182,6 +183,7 @@ export function fetchBaseQuery({
       headers = new Headers({}),
       body = undefined,
       params = undefined,
+      paramsOpts = undefined,
       responseHandler = 'json' as const,
       validateStatus = defaultValidateStatus,
       ...rest
@@ -216,7 +218,22 @@ export function fetchBaseQuery({
 
     if (params) {
       const divider = ~url.indexOf('?') ? '&' : '?'
-      const query = new URLSearchParams(stripUndefined(params))
+      let query = new URLSearchParams(stripUndefined(params))
+
+      if ('arrayFormat' in paramsOpts) {
+        switch (paramsOpts.arrayFormat) {
+          case 'repeat':
+            let temp = new URLSearchParams();
+            Object.entries(params).forEach(([k, v]) => {
+              if (Array.isArray(v)) {
+                v.forEach(o => temp.append(k, o))
+              } else temp.append(k, v)
+             })
+            query = temp;
+            break;
+        }
+      }
+
       url += divider + query
     }
 
```

**Comment by colemars** [comment]
Updated a few things

**Comment by StefanBRas** [comment]
Just for information; There is no standard way of doing this and it can be a [security risk](https://en.wikipedia.org/wiki/HTTP_parameter_pollution) if done in an unexpected way.

Of course RTKQ is mostly for frontend applications so security-wise it's more of a backend server issue, but you could still have problematic cases if you're not able to ensure that parsing a request and then using those parameters in RTKQ leads to the same request.

**Comment by colemars** [comment]
@StefanBRas Interesting. Out of curiosity, what is it about the `params` object that makes it less likely to be intercepted/manipulated than the `prepareParams` function?

**Comment by phryneas** [comment]
Hmm, honestly in that case I'd just leave it to the user to prepare a full querystring and pass it as an url, so something like

```js
myQuery: build.query({
  query(arg) {
    return {
      url: `/foo?${myPrepareSearchParams(arg)}`
    }
  }
})
```

Just because we have that `params` argument doesn't mean people *have* to use it and I'm not sure this is really inconvenient enough that we really need to add another escape hatch for it.

**Comment by colemars** [comment]
Yeah doesn't really support codegen users though atm. I was worried about decoupling from the provided `fetchBaseQuery` fn but I can just copy and patch in a hook.

**Comment by phryneas** [comment]
Hmm. I think it could be adressed in a `baseQuery`-wrapper that just directly converts the `param` to the `url` in that case?

**Comment by HansBrende** [comment]
> Hmm, honestly in that case I'd just leave it to the user to prepare a full querystring and pass it as an url, so something like
> 
> ```js
> myQuery: build.query({
>   query(arg) {
>     return {
>       url: `/foo?${myPrepareSearchParams(arg)}`
>     }
>   }
> })
> ```
> 
> Just because we have that `params` argument doesn't mean people _have_ to use it and I'm not sure this is really inconvenient enough that we really need to add another escape hatch for it.

@phryneas food for thought: if you want params to support both options, that's possible if you invert the behavior, i.e.:

as repeated items:
```typescript
const {data} = useXXXQuery({ids: [1, 2, 3]});
```
as a comma-delimited string:
```typescript
const {data} = useXXXQuery({ids: [1, 2, 3].join(',')})
```
The way it is now, only the comma-delimited string option can be supported via params since javascript objects cannot have multiples of the same key (but they can have a single key whose value is an array, obviously). So both of the 2 options above would essentially do the same thing with current behavior... and the _other_ option (the repeated item option) is impossible without circumventing the "normal" process. That seems suboptimal... two ways to do one desirable thing, zero ways to do the other desirable thing... maybe split the difference?

**Comment by phryneas** [comment]
All we do with `params` is pass it into `new URLSearchParams(params)`, so you can essentially do everything you could do there.

So even today, you could use the alternative notation of `[["foo", "1"], ["foo", "2"]]`. 

I don't really want to add any "nonstandard" other code on top. This is all [well documented in the MDN](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams).

As for shallow equality: if you do this on a component level, you'll have to use `useMemo` to keep it stable, but if you do this in `query` (where it should probably be done), you don't have to think about it a lot.

**Comment by HansBrende** [comment]
@phryneas I didn't realize the [["foo", "1"], ["foo", "2"]] syntax was supported by URLSearchParams. Good to know, thanks! Knowing that now: agreed.

As for shallow equality... how would we then go about having a multi-param along with some other query parameter?

For example, supposing I have 2 filter conditions:

```
useQuery({year: 2000, categories: ['books', 'movies']}
```
I'm not seeing any way to relay this thru to query without breaking the shallow equality functionality without some hack like `categories: ['books', 'movies'].join(',')`...

**Comment by phryneas** [comment]
If in doubt, memoize the object or put it into local component state.

In this case, the shortest solution might be
```js
useQuery(useMemo(() => {year: 2000, categories: ['books', 'movies']}, []))
```

**Comment by HansBrende** [comment]
@phryneas I hardcoded the strings in that particular example, but in all likelihood they'd be coming from the react state/props. So it would actually be:

```
useMemo(() => ({year, categories}), [year, categories]);
```
However, then we run into a problem: `['books', 'movies'] !== ['books', 'movies']` so react will refresh the memo even if the array elements stay the same.

Workaround:

```
useMemo(() => ({year, categories}), [year, ...categories]);
```
Now we run into another problem: React requires your dependency array to remain a constant length, which is not necessarily the case if the `categories` array is dynamically sized. If it is dynamically sized, React will only compare the first `min(array1.length, array2.length)` items causing a bug here.

2nd workaround:

```
useMemo(() => ({year, categories}), [year, categories.length, ...categories]);
```
This will work but you'll still get a nasty warning in dev console. Seems like a lot of work to get 1 multi-parameter to work correctly... why not just update shallowEquals to go one level deeper for array comparisons, to support url multi-params?

**Comment by phryneas** [comment]
Hm. To be honest, to this point it was "out of principle". In Redux, we never have gone for more than a shallow comparison and also React never does more than a shallow comparison, so it felt like the thing the ecosystem was already used to.

I can see the point where it is not optimal in edge cases and I'm willing to change my mind, but not without @markerikson's opinion on this and only if it really just is a very small bundle size increase.

We could also think about making the comparison function configurable, maybe on a `reactModule` level. Something like this :thinking: 
```js
const createApiWithDeepComparison = buildCreateApi(coreModule(), reactModule({ compareHookArgs: deepEquals}))
```

---

Internally, we are currently using a `useShallowStableValue` hook for this, so maybe adding a `useDeepStableValue` hook yourself in your codebase might be a good way to go until be have decided on that?

```ts
export function useShallowStableValue<T>(value: T) {
  const cache = useRef(value)
  useEffect(() => {
    if (!shallowEqual(cache.current, value)) {
      cache.current = value
    }
  }, [value])

  return shallowEqual(cache.current, value) ? cache.current : value
}
```

**Comment by phryneas** [comment]
Let's see what Mark thinks. For what it's worth, I know that Apollo does a deep equality check on the variables and it might just make sense to do that in general, at least for plain objects and arrays - I would probably want to skip any type of "class-like" thing with a warning there, as that should generally not be passed in as an argument into the query hooks (it will land in the store and cause warnings because of non-serializable values) there).

**Comment by markerikson** [comment]
So the specific proposal is for a 2-level check, then?

**Comment by HansBrende** [comment]
@markerikson yes, my specific proposal was for a 2-level depth for the shallow-equals comparison in order to support multi-params without resorting to horrible hacks. (And in my implementation above, I only start checking the second level if it is an array). 

However, if it makes more sense to do a full-depth comparison as @phryneas suggested (e.g. using fast-deep-equals or equivalent), that of course solves the use-case as well.

**Comment by TonisPiip** [comment]
In case someone else falls upon this problem.
Found two other solutions to allowing for array values in `parms`

First being 

```
Object.entries({
    foo:[1,2,3],
   bar:"baz",
                        }).flatMap(([key, values]) => Array.isArray(values) ? values.map((value) => [key, value]) : [[key, values]]).filter(v => ![undefined, ""].includes(v[1]))
```

the other using a package `qs`
`qs.stringify(queryArg, {arrayFormat: 'repeat'})`

see: https://www.npmjs.com/package/qs for details.

would be nice if there was a global way to overwrite how the searchquery string is formated from the params object.

**Comment by phryneas** [comment]
@TonisPiip that global way is `paramsSerializer`, which was mentioned directly in the the messages above yours.
